src/odm_sharing/private/queries.py

```diff
@@ -1,14 +1,326 @@
-from dataclasses import dataclass
-from typing import Dict
+import re
+from collections import defaultdict
+from dataclasses import dataclass, field
+from functools import partial
+from typing import Dict, List, Tuple
+# from pprint import pprint
 
+from functional import seq
+
+from odm_sharing.private.common import OrgName, TableName
 from odm_sharing.private.rules import RuleId
+from odm_sharing.private.stdext import StrEnum, sorted_dict
+from odm_sharing.private.utils import dqt
+
+from odm_sharing.private.trees import (
+    ALL_LIT,
+    Node,
+    NodeKind,
+    Op,
+    RangeKind,
+    RuleTree,
+    parse_op,
+)
+
+
+Args = List[str]
+Sql = str
 
 
-SqlQuery = str
+class SqlDialect(StrEnum):
+    OTHER = ''
+    MSSQL = 'mssql'
+    SYBASE = 'sybase'
 
 
 @dataclass(frozen=True)
 class Query:
-    data_sql: SqlQuery
-    rule_count_sqls: Dict[RuleId, SqlQuery]
-    column_sql: SqlQuery
+    sql: Sql
+    args: Args = field(default_factory=list)
+
+
+PartialQuery = Query  # incomplete query
+
+
+@dataclass(frozen=True)
+class TableQuery:
+    '''collection of queries for a single table'''
+    data_query: Query
+    rule_count_queries: Dict[RuleId, Query]
+    columns: List[str]  # pre-specified columns
+    select_rule_id: RuleId
+    _select_query: Query
+
+
+OrgTableQueries = Dict[OrgName, Dict[TableName, TableQuery]]
+
+
+INVALID_IDENT_PATTERN = re.compile(r'\W+', re.ASCII)
+
+
+def ident(x: str) -> str:
+    '''sanitize and quote sql identifier'''
+    return dqt(INVALID_IDENT_PATTERN.sub('', x))
+
+
+def convert(val: str) -> str:
+    '''convert to sql equivalent value'''
+    norm_val = val.lower()
+    if norm_val == 'true':
+        return '1'
+    elif norm_val == 'false':
+        return '0'
+    else:
+        return val
+
+
+def gen_sql_aux(
+    node: Node,
+    args: Args,
```

Comment:
this is a recursive function, and i need every recursion to append to the args list, so i think it would be difficult to write the recursion code if it's not passed in as an out-parameter.